E2_ESRGAN/lib/dataset.py

```diff
@@ -1,44 +1,114 @@
 import io
 from absl import logging
+from functools import partial
 import tensorflow as tf
 import tensorflow_datasets as tfds
 
 
-def scale_down(method="bicubic", dimension=1024, factor=4):
+def scale_down(method="bicubic", dimension=256, factor=4):
   def scale_fn(image, *args, **kwargs):
-    high_resolution = tf.image.resize_with_crop_or_pad(
-        image, dimension, dimension)
+    high_resolution = image
+    if not kwargs.get("no_random_crop", None):
+      high_resolution = tf.image.random_crop(
+          image, [dimension, dimension, image.shape[-1]])
+
     low_resolution = tf.image.resize(
-        image, [dimension // factor, dimension // factor], method=method)
-    return (low_resolution, high_resolution)
+        high_resolution,
+        [dimension // factor, dimension // factor],
+        method=method)
+    low_resolution = tf.clip_by_value(low_resolution, 0, 255)
+    high_resolution = tf.clip_by_value(high_resolution, 0, 255)
+    return low_resolution, high_resolution
+  scale_fn.dimension = dimension
   return scale_fn
 
 
+def augment_image(
+        low_res_map_fn,
+        brightness_delta=0.05,
+        contrast_factor=[0.7, 1.3],
+        saturation=[0.6, 1.6]):
+  """ helper function used for augmentation of images in the dataset. """
+  def augment_fn(low_resolution, high_resolution, *args, **kwargs):
+
+    if not tf.random.uniform(minval=0, maxval=5, shape=[], dtype=tf.int32):
+      # choosing integers from an uniform distribution on [0, 5)
+      # returns 0 with a probabilty of 20% (1 / (5 - 0) = 1/5)
+      # Roughly returning 20% of data without applying augmentation
+      return low_resolution, high_resolution
+
+    high_resolution = tf.image.rot90(
+        high_resolution, tf.random.uniform(
+            minval=1, maxval=5, dtype=tf.int32, shape=[]))
```

Comment:
Isn't `tf.image.rot90` the same as flip up/down for k=2? And same as no-op for k=4? Also, k=1 & k-5 will do the same thing. Instead:
1. Put minvalue = 1 and maxvalue=3
2. Remove `random_flip_up_down` from below.

Reply:
well, `tf.random.uniform` returns a value in range [minval, maxval)
So to get values through k=4, I need to put 5 in there.
but since you asked to put in that probabilty factor of 50%, I need to put minval=1 and maxval=4.